system/jlib/jmetrics.hpp

```diff
@@ -237,6 +236,101 @@ class CustomMetric : public MetricBase
 };
 
 
+template <typename T>
+struct BucketInfo
+{
+    std::string label;
+    T limit;
+};
+
+
+
+template <typename T>
+struct Bucket
+{
+    Bucket(const std::string &_label, T _limit) :
+        label{_label}, limit{_limit}, count{0} {}
+    std::string label;
+    T limit;
+    RelaxedAtomic<__uint64> count;
+};
+
+
+struct BucketValue
+{
+    BucketValue(const std::string &_label, __uint64 _count) : label(_label), count(_count) {}
+    std::string label;
+    __uint64 count = 0;
+};
+
+
+template <typename T>
+struct HistogramMetricValues
+{
+    std::vector<BucketValue> bucketValues;
+    __uint64 count;
+    T sum;
+};
+
+template<typename T>
+class jlib_decl HistogramMetric : public MetricBase
+{
+public:
+    HistogramMetric(const char *name, const char *desc, StatisticMeasure _units, const std::vector<BucketInfo<T>> &bucketDefs, const MetricMetaData &_metaData = MetricMetaData()) :
+        MetricBase(name, desc, MetricType::METRICS_HISTOGRAM, _units, _metaData)
+        {
+            for (auto &b: bucketDefs)
+            {
+                buckets.emplace_back(Bucket<T>{b.label, b.limit});
+            }
+        }
+
+    virtual __uint64 queryValue() const override
+    {
+        return count;
+    }
+
+    void recordMeasurement(T measurement)
+    {
+        count.fetch_add(1);           // total cumulative count of measurements
+        sum.fetch_add(measurement);   // sum of all measurements
+        getBucket(measurement).count.fetch_add(1);   // count by buckets
+    }
+
+    HistogramMetricValues<T> queryValues()
+    {
+        HistogramMetricValues<T> values;
+        values.count = count;
+        values.sum = sum;
+        for (auto const &bucket: buckets)
+        {
+            values.bucketValues.emplace_back(BucketValue{bucket.label, bucket.count});
```

Comment:
Does the result need to include the label?  This should be available to the sink from the meta.

Reply:
Sorry for the confusion, I had forgotten what the meta data was used for.

Reporting the metrics isn't called very often, so the efficiency is not that important, but cloning the string labels each time is less efficient.